b2sdk/file_version.py

```diff
@@ -488,23 +488,55 @@ def from_api_response(self, file_version_dict, force_action=None):
                "replicationStatus": "completed"
            }
 
+        or this:
+
+        .. code-block:: python
+
+            {
+                "content-length": 11, 
+                "content-type": "b2/x-auto", 
+                "x-bz-content-sha1": "2aae6c35c94fcfb415dbe95f408b9ce91ee846ed", 
+                "x-bz-upload-timestamp": 5000, 
+                "x-bz-file-id": "9999", 
+                "x-bz-file-name": "file"
+            }
+
```

Comment:
@agoncharov-reef can you give me some more guidance on this? why do you think it should be two methods initially? ho would you call them? I am thinking about separating this into 3 methods, as per your suggestion, but as you can imagine `from_api_response` is a method that is called rather often :)

Reply:
I don't think you should untangle whatever was there before you, but maybe not tangle it even more - that's what I meant. And please correct me if I'm wrong below.

I see that `B2Session.get_file_info_by_name()` calls `B2RawHTTPApi` which generates download url and makes HEAD request to b2. So `B2Session.get_file_info_by_name` will always return headers from that download HEAD request, and nothing more:
```python
    def get_file_info_by_name(self, bucket_name: str, file_name: str) -> FileVersion:
        return self.file_version_factory.from_api_response(
            self.session.get_file_info_by_name(bucket_name, file_name),  <-- this is a response from HEAD
            force_action='folder',
        )
```

Compare it to `B2Session.get_file_info_by_id()` which under the hood makes a fair POST request and gets all information, not only headers:
```python
    def get_file_info(self, file_id: str) -> FileVersion:
        return self.file_version_factory.from_api_response(
            self.session.get_file_info_by_id(file_id)   <-- this is a response from POST
        )
```

So I thought that maybe if there are different types of data then you shouldn't process them with the same function. And my initial proposal was to create a `file_version_factory.from_api_head_response` method which would be able to create `FileVersion` just from headers information and nothing more, while `file_version_factory.from_api_response` would work with all the data from fair POST request.

---

However, during writing this I found that there is also `download_version_factory` in `B2Api` class:
```python
        self.file_version_factory = self.FILE_VERSION_FACTORY_CLASS(self)
        self.download_version_factory = self.DOWNLOAD_VERSION_FACTORY_CLASS(self)
```

I was wondering what it is, and here is its definition:
```python
class DownloadVersionFactory:
    """
    Construct :py:class:`b2sdk.v2.DownloadVersion` objects from download headers.
    """
```
But as I understand, "download headers" is exactly what we get when we call `B2Session.get_file_info_by_name()`. So if my assumption is correct, your `B2Api.get_file_info_by_name` could now look like:
```python
    def get_file_info_by_name(self, bucket_name: str, file_name: str) -> DownloadVersion:
        return self.download_version_factory.from_response_headers(
            self.session.get_file_info_by_name(bucket_name, file_name),
        )
```
However, now it returns `DownloadVersion` instead of file version. I wanted to solve this, so I searched in code and found `file_version_info_from_download_version` which converts `DownloadVersion` to `FileVersion` (and it also answers your question about `action`):
```python
def file_version_info_from_download_version(download_version: v2.DownloadVersion):
    return FileVersionInfo(
        id_=download_version.id_,
        file_name=download_version.file_name,
        size=download_version.size,
        content_type=download_version.content_type,
        content_sha1=download_version.content_sha1,
        file_info=download_version.file_info,
        upload_timestamp=download_version.upload_timestamp,
        action='upload',
        content_md5=None,
        server_side_encryption=download_version.server_side_encryption,
        file_retention=download_version.file_retention,
        legal_hold=download_version.legal_hold,
        cache_control=download_version.cache_control,
        api=download_version.api,
    )
```

So now your code could look something like this:
```python
    def get_file_info_by_name(self, bucket_name: str, file_name: str) -> FileVersion:
        download_version = self.download_version_factory.from_response_headers(
            self.session.get_file_info_by_name(bucket_name, file_name),
        )
        return file_version_info_from_download_version(download_version)
```

---

I was curious where `file_version_info_from_download_version` is used, and I found this in `v1/bucket.py::Bucket
```python
class Bucket:
    ...
    def get_file_info_by_name(self, file_name: str) -> FileVersionInfo:
        return file_version_info_from_download_version(super().get_file_info_by_name(file_name))
```

And that `super().get_file_info_by_name(...)` is this:
```python
    def get_file_info_by_name(self, file_name: str) -> DownloadVersion:
        """
        Gets a file's DownloadVersion by name.

        :param str file_name: the name of the file who's info will be retrieved.
        """
        try:
            return self.api.download_version_factory.from_response_headers(
                self.api.session.get_file_info_by_name(self.name, file_name)
            )
        except FileOrBucketNotFound:
            raise FileNotPresent(bucket_name=self.name, file_id_or_name=file_name)
```

And all this is exactly what I proposed: get headers from `get_file_info_by_name`, convert them to `DownloadVersion`, then convert it to `FileVersion`.

So it looks like `v1.Bucket.get_file_info_by_name()` returns `FileVersionInfo`, while latest `Bucket.get_file_info_by_name()` returns `DownloadVersion`. I think the latter is more correct, since you cannot recreate full file version from headers; and it's newer anyway. So let's use it!

```python
class B2Api:
    ...
    def get_file_info_by_name(self, bucket_name: str, file_name: str) -> DownloadVersion:
        bucket = self.get_bucket_by_name(bucket_name)
        return bucket.get_file_info_by_name(file_name)
```

Reply:
thanks a lot for this comprehensive review. I reverted back to the previous version 'file_info.py` and now I am using `Bucket.get_file_info_by_name()` to pull the information. there are much less changes in the code currently